File: Sources/HarmonyTrainer/Exercise.swift
import Foundation

// MARK: - ExerciseMode

public enum ExerciseMode: String, Codable, Equatable, Hashable {
    case freePlay              // Detect and label what the user plays
    case chordPrompt           // "Play the IV chord in C major"
    case progressionGuide      // Step through a progression
    case scalePractice         // Play a scale with degree feedback
}

// MARK: - ExercisePrompt

public struct ExercisePrompt: Equatable, Codable, Hashable {
    public let instruction: String
    public let expectedPitchClasses: Set<Int>
    public let acceptInversions: Bool
    public let degree: Int?
    public let romanNumeral: String?

    public init(
        instruction: String,
        expectedPitchClasses: Set<Int>,
        acceptInversions: Bool = true,
        degree: Int? = nil,
        romanNumeral: String? = nil
    ) {
        self.instruction = instruction
        self.expectedPitchClasses = expectedPitchClasses
        self.acceptInversions = acceptInversions
        self.degree = degree
        self.romanNumeral = romanNumeral
    }
}

// MARK: - ExerciseResult

public enum ExerciseResult: String, Codable, Equatable, Hashable {
    case correct
    case incorrect
    case partial        // Some notes right, not all
    case waiting        // Not enough notes yet
}

// MARK: - ExerciseState

public struct ExerciseState: Equatable, Codable, Hashable {
    public var mode: ExerciseMode
    public var currentPrompt: ExercisePrompt?
    public var result: ExerciseResult
    public var progressionTemplate: ProgressionTemplate?
    public var progressionIndex: Int
    public var totalSteps: Int
    public var completedCount: Int
    public var streak: Int

    public init(mode: ExerciseMode = .freePlay) {
        self.mode = mode
        self.currentPrompt = nil
        self.result = .waiting
        self.progressionTemplate = nil
        self.progressionIndex = 0
        self.totalSteps = 0
        self.completedCount = 0
        self.streak = 0
    }
}

// MARK: - Exercise Evaluation

public enum ExerciseEvaluation {
    /// Evaluate active pitch classes against an exercise prompt.
    public static func evaluate(
        activePitchClasses: Set<Int>,
        prompt: ExercisePrompt
    ) -> ExerciseResult {
        guard !prompt.expectedPitchClasses.isEmpty else { return .waiting }

        if activePitchClasses.isEmpty {
            return .waiting
        }

        if prompt.acceptInversions {
            // Match pitch classes regardless of octave/inversion
            if activePitchClasses == prompt.expectedPitchClasses {
                return .correct
            }
            if activePitchClasses.isSubset(of: prompt.expectedPitchClasses) && activePitchClasses.count >= 2 {
                return .partial
            }
            if activePitchClasses.count >= prompt.expectedPitchClasses.count {
                return .incorrect
            }
            return .waiting
        } else {
            // Strict: exact match required
            if activePitchClasses == prompt.expectedPitchClasses {
                return .correct
            }
            if activePitchClasses.isSubset(of: prompt.expectedPitchClasses) {
                return .partial
            }
            return .incorrect
        }
    }
}

File: web/src/domain/exercise.ts
import type { ProgressionTemplate } from './progressionTemplate.js';

export enum ExerciseMode {
  FreePlay = 'freePlay',
  ChordPrompt = 'chordPrompt',
  ProgressionGuide = 'progressionGuide',
  ScalePractice = 'scalePractice',
}

export interface ExercisePrompt {
  readonly instruction: string;
  readonly expectedPitchClasses: ReadonlySet<number>;
  readonly acceptInversions: boolean;
  readonly degree?: number;
  readonly romanNumeral?: string;
}

export enum ExerciseResult {
  Correct = 'correct',
  Incorrect = 'incorrect',
  Partial = 'partial',
  Waiting = 'waiting',
}

export interface ExerciseState {
  mode: ExerciseMode;
  currentPrompt: ExercisePrompt | null;
  result: ExerciseResult;
  progressionTemplate: ProgressionTemplate | null;
  progressionIndex: number;
  totalSteps: number;
  completedCount: number;
  streak: number;
}

export function createExerciseState(mode: ExerciseMode = ExerciseMode.FreePlay): ExerciseState {
  return {
    mode,
    currentPrompt: null,
    result: ExerciseResult.Waiting,
    progressionTemplate: null,
    progressionIndex: 0,
    totalSteps: 0,
    completedCount: 0,
    streak: 0,
  };
}

export function evaluate(
  activePitchClasses: ReadonlySet<number>,
  prompt: ExercisePrompt,
): ExerciseResult {
  if (prompt.expectedPitchClasses.size === 0) return ExerciseResult.Waiting;
  if (activePitchClasses.size === 0) return ExerciseResult.Waiting;

  if (prompt.acceptInversions) {
    if (setsEqual(activePitchClasses, prompt.expectedPitchClasses)) {
      return ExerciseResult.Correct;
    }
    if (isSubset(activePitchClasses, prompt.expectedPitchClasses) && activePitchClasses.size >= 2) {
      return ExerciseResult.Partial;
    }
    if (activePitchClasses.size >= prompt.expectedPitchClasses.size) {
      return ExerciseResult.Incorrect;
    }
    return ExerciseResult.Waiting;
  } else {
    if (setsEqual(activePitchClasses, prompt.expectedPitchClasses)) {
      return ExerciseResult.Correct;
    }
    if (isSubset(activePitchClasses, prompt.expectedPitchClasses)) {
      return ExerciseResult.Partial;
    }
    return ExerciseResult.Incorrect;
  }
}

function setsEqual(a: ReadonlySet<number>, b: ReadonlySet<number>): boolean {
  if (a.size !== b.size) return false;
  for (const v of a) {
    if (!b.has(v)) return false;
  }
  return true;
}

function isSubset(sub: ReadonlySet<number>, sup: ReadonlySet<number>): boolean {
  for (const v of sub) {
    if (!sup.has(v)) return false;
  }
  return true;
}
